SPEC.md

```diff
@@ -33,69 +33,35 @@ Upon completion of the container lifecycle, the runtime executes the plugins in
+- `CNI_NETNS`: Path to network namespace file representing network namespace to which the container is to be added, i.e. /proc/[pid]/ns/net or a bind-mount/link to it.
+- `CNI_IFNAME`: Interface name to set up. This is the name that should be assigned to the interface created inside the container (network namespace); consequently it must comply with the standard Linux restrictions on interface names.
+- `CNI_ARGS`: Extra arguments passed in by the user at invocation time. Alphanumeric key-value pairs separated by semicolons; for example, "FOO=BAR;ABC=123". This allows granular configuration of CNI plugins on a per-container basis.
+- `CNI_PATH`: Colon-separated list of paths to search for CNI plugin executables.
 
-### Result
+#### Network configuration streamed via standard input
+This is a JSON document describing a network to which a container is to be joined.
+The schema is described below.
 
-Success is indicated by a return code of zero and the following JSON printed to stdout in the case of the ADD command. This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details).
-
-```
-{
-  "ip4": {
-    "ip": <ipv4-and-subnet-in-CIDR>,
-    "gateway": <ipv4-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv4-routes>    (optional)
-  },
-  "ip6": {
-    "ip": <ipv6-and-subnet-in-CIDR>,
-    "gateway": <ipv6-of-the-gateway>,  (optional)
-    "routes": <list-of-ipv6-routes>    (optional)
-  },
-  "dns": <list-of-DNS-nameservers>     (optional)
-}
-```
-
-The "dns" field contains a list of a priority-ordered list of DNS nameservers that this network is aware of.
-Each entry in the list is a string containing either an IPv4 or an IPv6 address.
-Typically this value would just be the value returned by the IPAM plugin.
-It is outside the scope of this specification how the container runtime uses the list of DNS nameservers from each of the networks to provide name resolution services to the container.
-Examples of how this list could be used include generating an `/etc/resolv.conf` file to be injected into the container filesystem or running a DNS forwarder on the host.
+#### Result streamed via standard output
+Success is indicated by an exit code of zero and a JSON result printed to stdout.
+This should be the same output as was returned by the IPAM plugin (see [IP Allocation](#ip-allocation) for details) with the exception of `state` field.
+`state` field should contain both the state that the plugin itself needs the runtime to persist as well as any `state` returned by the IPAM plugin.
```

Comment:
You're correct that it's up to the plugin to amalgamate. For example, if the plugin does not want to keep any state, it could just pass IPAM's state up as its own.

As for specifying it rules or conventions on how to combine state is an option. If we go down that route, I'd like to keep it generic as I would also like to support more arbitrary plugin composability. For example, one plugin might call onto 2 helpers. That's why I decided not to prescribe this.